rotkehlchen/chain/evm/manager.py

```diff
@@ -0,0 +1,1220 @@
+            new_events_web3[e_idx]['topics'] = new_topics
+            new_events_web3[e_idx]['transactionHash'] = event['transactionHash'].hex()
+
+        start_block = end_block + 1
+        events.extend(new_events_web3)
+        # end of the loop, end of 1 query. Reset the block range to max
+        block_range = initial_block_range
+
+    return events
+
+
+class EvmManager(LockableQueryMixIn, metaclass=ABCMeta):
+    """EvmManager defines a basic implementation for EVM chains.
+
+    The child class must implement the following methods:
+    - query_highest_block
+    - have_archive
+    """
+    def __init__(
+            self,
+            etherscan: Etherscan,
+            msg_aggregator: MessagesAggregator,
+            greenlet_manager: GreenletManager,
+            connect_at_start: Sequence[WeightedNode],
+            database: 'DBHandler',
+            etherscan_node: WeightedNode,
+            etherscan_node_name: str,
+            blockchain: SupportedBlockchain,
+            contract_scan: EvmContract,
+            contract_multicall: EvmContract,
+            contract_multicall_2: EvmContract,
+            graph_url: str,
+            rpc_timeout: int = DEFAULT_EVM_RPC_TIMEOUT,
+    ) -> None:
+        super().__init__()
+        self.greenlet_manager = greenlet_manager
+        self.web3_mapping: Dict[NodeName, Web3] = {}
+        self.etherscan = etherscan
+        self.msg_aggregator = msg_aggregator
+        self.database = database
+        self.etherscan_node = etherscan_node
+        self.etherscan_node_name = etherscan_node_name
+        self.blockchain = blockchain
+        self.chain_id = SUPPORTED_BLOCKCHAIN_TO_CHAINID[blockchain]
+        self.chain_name = self.chain_id.serialize()
+        self.contract_scan = contract_scan
+        self.contract_multicall = contract_multicall
+        self.contract_multicall_2 = contract_multicall_2
+        self.rpc_timeout = rpc_timeout
+        self.archive_connection = False
+        self.queried_archive_connection = False
+        self.connect_to_multiple_nodes(connect_at_start)
+        self.blocks_subgraph = Graph(graph_url)
+
+        # A cache for the erc20 contract info to not requery same one
+        self.contract_info_cache: Dict[ChecksumEvmAddress, Dict[str, Any]] = {}
+
+    def get_multi_balance(
+            self,
+            accounts: List[ChecksumEvmAddress],
+            call_order: Optional[Sequence[WeightedNode]] = None,
+    ) -> Dict[ChecksumEvmAddress, FVal]:
+        """Returns a dict with keys being accounts and balances in ETH
+
+        May raise:
+        - RemoteError if an external service such as Etherscan is queried and
+          there is a problem with its query.
+        """
+        balances: Dict[ChecksumEvmAddress, FVal] = {}
+        log.debug(
+            'Querying {self.chain_name} chain for {self.blockchain.value} balance',
+            eth_addresses=accounts,
+        )
+        result = self.contract_scan.call(
```

Comment:
this is always making the assumption that a scan contract exists. We need to probably modify those